### ISSUE TITLE
Check out for ESFM emulation

### ISSUE THREAD

**Issue by Wohlstand** [Wohlstand/OPL3BankEditor#164]
ESFM is an extended clone of OPL3. Recently I got the ESS card with ES1688F chip that has ESFM. I tried it in action with my old Windows 98 machine, and it's very interesting thing. I heard about ESFM before from @jpcima, however, I had't much information about it and how it actually sounds. So, getting the real ESS card with a compatible chip gave me some proveground for researches.

> ESS Technology's in-house developed derivative, termed ESFM, is an enhanced, 20-voice, 72-operator, OPL3-compatible clone incorporating two operating modes, a Native mode and a Legacy mode, which controls its feature-set and behavior. In Native mode, ESFM allows more than six 4-operator FM voices to be mapped, potentially allowing for a significant increase in the complexity of tones generated. The drivers for Windows 9x incorporate their own custom instrument patches which make use of this extended mode. Conversely, Legacy mode provides full backward-compatibility with Yamaha's YMF262. ESFM's output in this mode is moderately faithful to the YMF262 overall, but some tones are rendered quite differently, resulting in unique distortions in the sound and music of some games.

Some details at VOGONS: https://www.vogons.org/viewtopic.php?f=46&t=37497&start=120#p738552

**Comment by freq-mod** [comment]
Yeah, interesting card. In its nativ mode, it reportedly has 18 4-op channels (18x4=72). But it's not a dual OPL3, so 2-op channels count stays the same.

Reverse enginnering project: https://github.com/pachuco/ESSPlayMid

**Comment by pachuco** [comment]
https://github.com/pachuco/Buttler
I'm working on separate project of providing hardware read/write port access to other people online, so people can remotely play with other people's hardware. Which is totes not a reckless and dangerous thing to do.
I'll lend my ESSFM for study on request.

Currently not finished, so stay tuned, I guess.

**Comment by sneakernets** [comment]
This is a great idea. I would love to write/extend DMXOPL3 to support this if there are more features.

**Comment by freq-mod** [comment]
ESFM is getting properly investigated. Turns out, it's **MUCH**  more capable... https://github.com/jwt27/esfm

Per-op frequency, customizable op connections, per 2-op pair envelope, customizable LFO, noise generation (?), ability to output sound and modulate another op at once, and the list goes on.

**Comment by Wohlstand** [comment]
Very interesting, I should take a look on that soon 👀

**Comment by freq-mod** [comment]
https://discord.gg/C9FRkma5  - Discord server for ESFM research.

**Comment by leecher1337** [comment]
In case someone wants to enhance the original sound card driver, I reverse engineered the Windows 2000 driver (which uses ESFM mode, if supported by card)  to build a 64bit Windows driver for the ES1969 card: https://github.com/leecher1337/es1969
May still have some bugs, but generally works on my ES1969.

**Comment by Wohlstand** [comment]
Adding support for new bank format isn't hard: you need to just make a class that inherits the https://github.com/Wohlstand/OPL3BankEditor/blob/master/src/FileFormats/ffmt_base.h, and use existing files as an example (take the apogee TMB code as a simplest example). And then, add it into the ffmt_factory.cpp to be registered and being available.

However, if speak about adding new features generally, that yeah, needs a modify the structure of bank and instruments, and additionally implement missing UI fields.

Anyway, I'll try to take a look by myself soon.

**Comment by leecher1337** [comment]
Yeah, figured that out, but adding only format definitions is useless in this case.

**Comment by Wohlstand** [comment]
The only last case is left - the emulation-side support to let my thing to preview these instruments.

**Comment by leecher1337** [comment]
Don't know how far https://github.com/Kagamiin/ESFMu progressed with emulation.

**Comment by freq-mod** [comment]
ESFMu emulator is fairly complete, except cymbal noise pattern emulation I tink (yk, ESFM does some things differently than OPL3). It will get integrated into Furnace for the next release.

**Comment by Wohlstand** [comment]
> Don't know how far https://github.com/Kagamiin/ESFMu progressed with emulation.

Few moments ago I added it into libADLMIDI for some tests, very soon I will add into OPL3 Bank Editor too. And for the first time it will work as one another OPL3 chip emulator until the ESFM functionality will be handled later.

**Comment by freq-mod** [comment]
Keep in mind ESFM has far more parameters than OPL3:
each operator can be different pitch (both in cents and semitones), which can be in theory abused for 72ch
there is envelope delay parameter this seems to enable operator later than others
deep vibrato and deep tremolo are per-operator
there is no concept of algorithm either. operator connection and carrier/modulator is controlled by mod input and output level parameters, which are 3-bit
panning is also per-operator
noise mode works differently, rather than stealing 3ch into 5 drum tones it takes two last operators and performs its special drum synthesis on them
there is no feedback (which internally works vastly differently than on OPL), use modinput on operator 1 to emulate it
and finally, it can go well lower than MIDI note #0

**Comment by Wohlstand** [comment]
> Keep in mind ESFM has far more parameters than OPL3:

Ye, I know that, and I going to learn that in details, especially because I am required to implement this support.

Thanks for the hints!

**Comment by freq-mod** [comment]
For even more hints: https://github.com/tildearrow/furnace/blob/master/doc/4-instrument/fm-esfm.md

> P.S. I got another sample of ESFM sound card, with a P3 PC suddenly gifted by my neighbour, and I repaired it, and I found it has the ESFM chip onboard. I also could try to support the hardware too in addition to the emulator.

Wow, great news! Wish more OPL music software allowed for real hardware passthrough....

**Comment by freq-mod** [comment]
![Image](https://github.com/user-attachments/assets/1bcd8182-3336-4da4-bf49-2848821b8899)

I was thinking about the layout of parameters, it is was to be implemented in OPL3BE, and I came out with something like this. With some adjustments, it should fit all four operators on this window.

not featured is normal vibrato and tremolo, and fixed frequency mode (where you input FreqNum and Octave block numbers manually)

**Comment by Kagamiin** [comment]
(K-s) Hey, nice to see ESFM being implemented into OPL3BankEditor, @Wohlstand! x3

If you need any help with stuff, be sure to contact me or the ESFM research team in our Discord server. x3

Also yeah we need people experienced with reading chip decaps in order to smooth out the last couple issues with ESFMu. Other than the math edge cases when you abuse chained modulation, and the cymbal noise mode weirdness, it's pretty much 100% accurate to the ear tho.

**Comment by freq-mod** [comment]
@Wohlstand https://github.com/tgies/ESFMu - faster ESFMu. Up to 2x speed improvement, same output.

@tgies please consider PRing it here too

**Comment by Wohlstand** [comment]
> [@Wohlstand](https://github.com/Wohlstand) https://github.com/tgies/ESFMu - faster ESFMu. Up to 2x speed improvement, same output.

Sounds interesting, I definitely should try it out. Accordign my last benchmark, before fixes it worked 2.5x slower than regular Nuked, and on my phone it lagged even being a single chip. I should check how hard it will work on my phone after update.

**Comment by tgies** [comment]
> [@Wohlstand](https://github.com/Wohlstand) https://github.com/tgies/ESFMu - faster ESFMu. Up to 2x speed improvement, same output.
> 
> [@tgies](https://github.com/tgies) please consider PRing it here too

I have a PR to upstream (kagamiin/ESFMu#20) to see if they will pick up these changes; I will hold off on PRing my fork anywhere until I see what happens with that.

Nuked-OPL3-fast is only maintained as a fork because nukeykt didn't want to upstream such a big rewrite. I'd rather not "usurp" everyone working on FM emulation and get myself on the hook for maintaining a bunch of forks in the process.

**Comment by Wohlstand** [comment]
Primarily he needs to have sure he knows the code well and nothing unexpected and uncontrollable appears. And ye, to keep an ability to receive updates for both, and just backport them between both emulatos I keeping both original and optimised. However, initially I kept optimised separately since it was an experimental while original is stable.

**Comment by Kagamiin** [comment]
> I have a PR to upstream ([Kagamiin/ESFMu#20](https://github.com/Kagamiin/ESFMu/pull/20)) to see if they will pick up these changes; I will hold off on PRing my fork anywhere until I see what happens with that.

I have received the PR this morning and I'm taking a look at the changes. Unfortunately, due to personal reasons I'm struggling to dedicate much time to personal/community projects, so I apologize if I take too long to finish reviewing it.

**Comment by freq-mod** [comment]
ESFMu-fast has been merged. It can be backported now
